content/en/blog/2024/llm-observability.md

```diff
@@ -0,0 +1,229 @@
+
+In the following example, we'll use [Prometheus](https://prometheus.io/) and
+[Grafana Tempo](https://grafana.com/oss/tempo/) as the target backend for
+metrics and traces generated by an auto-instrumentation LLM monitoring library
+[OpenLIT](https://github.com/openlit/openlit). We will use
+[Grafana](https://grafana.com/oss/grafana/) as the tool to visualize the LLM
+monitoring data. You can choose any backend of your choice to store OTel metrics
+and traces.
+
+## Why Observability Matters for LLM Applications
+
+Monitoring LLM applications is crucial for several reasons.
+
+1. It's vital to keep track of how often LLMs are being used, especially in
+   RAG-based applications. This monitoring helps you understand the demand and
+   usage patterns.
+2. Understanding how long it takes for LLMs to respond to requests and the
+   nature of each interaction (prompt and response) can significantly impact
+   user experience and operational efficiency. Since using these models incurs
+   costs per request, tracking usage helps in managing expenses.
+3. Making too many requests in a short timeframe can lead to rate limiting,
+   temporarily blocking further requests.
+
+By keeping a close eye on these aspects, you can not only save costs but also
+avoid hitting request limits, ensuring your LLM applications perform optimally.
+
+## What are the signal that you should be looking at?
+
+Integrating Large Language Models (LLMs) into applications represents a pivotal
+shift from traditional machine learning (ML) practices, particularly because
+these LLMs are accessed via API calls rather than being hosted locally or run
+in-house. This external integration necessitates a nuanced approach to
+observability, where understanding both the sequence of events (through traces)
+and the aggregated data (through metrics) becomes crucial for optimizing
+performance and managing costs. Here's the critical signals to monitor:
+
+### Traces
+
+Tracing is vital for capturing the flow of requests through your system and
+understanding the interactions with hosted LLM services. Here's what you need to
+focus on:
+
+- **Request Metadata**: This is even more critical in the context of LLMs than
+  traditional ML, given the variety of parameters (like temperature and prompt
+  details) that can drastically affect both the response quality and the cost.
+  Specific aspects to monitor are:
+  - **Temperature**: Indicates the level of creativity or randomness desired
+    from the model's outputs. Varying this parameter can significantly impact
+    the nature of the generated content.
+  - **Model Name or Version**: Essential for tracking over time, as updates to
+    the LLM might affect performance or response characteristics.
+  - **Prompt Details**: The exact inputs sent to the LLM, which, unlike in-house
+    ML models where inputs might be more controlled and homogeneous, can vary
+    wildly and affect output complexity and cost implications.
```

Comment:
Note that in the LLM Working Group, we're explicitly not capturing these details on spans because many backend systems can struggle with those often large payloads.